broker/README.md

```diff
@@ -0,0 +1,287 @@
+- may _not_ contain capital letters
+- may _not_ contain any other characters including spaces, "-", "_", "/", etc.
+
+<!-- fe Create, deploy, and utilize a testing instance of the broker -->
+
+### Setup a _production_ instance of the broker
+<!-- fs -->
+Strictly speaking, this tutorial is not intended to cover this topic, but the only difference between a testing and production instance of the broker is the presence/absence of a tag (string) appended to the names of all GCP resources that created for, and used by, the broker instance.
+So, I will cover some relevant items here.
+
+The production instance only needs to be _setup_ once and it has already been done.
+
+Most components of the broker can be _updated_ by simply replacing the relevant files in the `broker-files` GCS bucket, which has been set up to stage the files used to run the nightly broker.
+This can be done manually, or by re-running the setup using `setup_broker.sh`.
+By design, the setup script(s) will skip the _creation_ of any resources that already exist, but it will still upload new broker files to the staging bucket, overwriting any existing files.
+
+To __setup the production broker__ the process is exactly the same as for a testing instance, except that we set the testid to `False`.
+This builds and uses GCP resources that do _not_ have a testid appended to the names.
+
+The "teardown" options that can be triggered in the setup scripts are specifically configured to _prevent_ a user from deleting _production_ resources, even if the user intentionally tries to do so.
+This is one layer of security, but we should still look into protecting those resources in other ways (e.g., make backups of databases and buckets, find out which resources can be configured directly to prevent deletion and configure them).
+
+<!-- fe Create and deploy a production instance of the broker -->
+
+
+## Detailed Workflow with code examples
+<!-- fs -->
+To test the broker, you should perform these 3 steps:
+1. __Setup__ your own instance of the broker for testing.
+2. __Run your tests__ using the instance you created in step 1.
+3. __Teardown__ the broker instance when you are done (i.e., delete the GCP resources created in step 1)
+
+You can also leave the testing instance inactive, just be sure to shutdown all VMs and end all Dataflow jobs.
+
+### Setup a testing instance
+<!-- fs -->
+If you don't have GCP command line tools like `gcloud` and `gsutil` installed, follow the instructions in step 2 of [../README.md](../README.md#setup-the-broker-for-the-first-time).
+Activate your virtual environment if needed.
+
+```bash
+#--- Set the project ID env variable if it's not already
+export GOOGLE_CLOUD_PROJECT=ardent-cycling-243415
+
+#--- The Compute Engine VMs must be assigned to a specific zone.
+# We currently use the same zone for all instances.
+# The default is us-central1-a,
+# but it can controlled explicitly by setting an environment variable
+export CE_zone=us-central1-a
```

Comment:
Conventionally environment variables are all caps.  Is there a reason `CE_ZONE` is not?

Reply:
No good reason, I just didn't know that. I'll change it.

Reply:
The code comment implies that `CE_ZONE` is used by GCP and friends.  But am I correct that `CE_ZONE` is just used here in this file down below?

Reply:
Currently that is probably correct, that is a detail I haven't quite finished. The `zone` is still hardcoded in some places, but my intention is to change that so that setting it affects all things you would hope/expect it to affect. It's on my list.